odl/solvers/functional/default_functionals.py

```diff
@@ -0,0 +1,731 @@
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Default functionals defined on any (reasonable) space."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+
+from odl.solvers.functional.functional import Functional
+from odl.operator.operator import Operator
+from odl.solvers.advanced.proximal_operators import (
+    proximal_l1, proximal_cconj_l1, proximal_l2, proximal_cconj_l2,
+    proximal_l2_squared)
+
+from odl.operator.default_ops import (ZeroOperator, IdentityOperator)
+
+
+__all__ = ('L1Norm', 'L2Norm', 'L2NormSquare', 'ZeroFunctional',
+           'ConstantFunctional')
+
+
+# TODO: Implement some of the missing gradients
+
+class L1Norm(Functional):
+
+    """The functional corresponding to L1-norm.
+
+    The L1-norm, ``||x||_1``,  is defined as the integral/sum of ``|x|``.
+
+    References
+    ----------
+    Wikipedia article on `Norms
+        <https://en.wikipedia.org/wiki/Norm_(mathematics)>`_.
+    """
+
+    def __init__(self, domain):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        domain : `LinearSpace`
+            Domain of the functional.
+        """
+        super().__init__(domain=domain, linear=False, grad_lipschitz=np.inf)
+
+    def _call(self, x):
+        """Applies the functional to the given point.
+
+        Parameters
+        ----------
+        x : `LinearSpaceVector`
+            Element in the domain of the functional.
+
+        Returns
+        -------
+        out : `field` element
+            Result of the evaluation.
+        """
+        return np.abs(x).inner(self.domain.one())
+
+    @property
+    def gradient(x):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        # It does not exist on all of the space
+        raise NotImplementedError('functional not differential.')
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Step size-like parameter
+
+        Returns
+        -------
+        out : Operator
+            Domain and range equal to domain of functional
+        """
+
+        return proximal_l1(space=self.domain)(sigma)
+
+    @property
+    def conjugate_functional(self):
+        """The convex conjugate functional of the L1-norm."""
+        return IndicatorLinftyUnitBall(self.domain)
+
+
+class IndicatorLinftyUnitBall(Functional):
```

Comment:
`np.sqrt(x.inner(x))` always equals `x.norm()**2` so it doesnt add much. This could all be improved by the pesky integral thing. We really should add it.